### PR TITLE
Increase Groovy version to 3.0.6 and give condition closures access to the specification instance

### LINKED ISSUE #1177: Using thrown() causes IllegalAccessError for 2.0-M2-groovy-3.0 on JDK 8 with noverify switch
# Issue description
Not really sure if this needs fixing, but I'm reporting anyway so you can decide.

When running Spock 2.0-M2 with Groovy 3 on JDK 8 with `-noverify` switch, I'm getting `llegalAccessError` on any attempt of using `thrown()` method.
There is no problem with JDK 11, and with JDK 8 there is no error if `-noverify` is removed.

I usually run my tests with `-noverify` and `-XX:TieredStopAtLevel=1`options set. This is inspired by numerous talks and blogs from Spring's Dave Syer (i.e. https://spring.io/blog/2018/12/12/how-fast-is-spring).

# How to reproduce
Use JDK 8. With Gradle the following in `build.gradle` for tests
```
  test {
    jvmArgs = ["-noverify", "-XX:TieredStopAtLevel=1"]
    useJUnitPlatform()
  }
```

Execute any test containing any variant of `thrown()`.

# Additional Environment information
## Java/JDK
```
java -version
java version "1.8.0_251"
Java(TM) SE Runtime Environment (build 1.8.0_251-b08)
Java HotSpot(TM) 64-Bit Server VM (build 25.251-b08, mixed mode)
```

## Groovy version
3.0.4

### Gradle
```
gw --version
------------------------------------------------------------
Gradle 6.4.1
------------------------------------------------------------

Build time:   2020-05-15 19:43:40 UTC
Revision:     1a04183c502614b5c80e33d603074e0b4a2777c5

Kotlin:       1.3.71
Groovy:       2.5.10
Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
JVM:          11.0.7 (Oracle Corporation 11.0.7+8-LTS)
OS:           Mac OS X 10.15.5 x86_64
```

## Operating System
 Mac OS X 10.15.5 x86_64

### Gradle/Grails
`testImplementation "org.spockframework:spock-core:2.0-M2-groovy-3.0"`

### PR DESCRIPTION
**Pull request by Vampire**
fixes #1177

---
This change is [Reviewable](https://reviewable.io/reviews/spockframework/spock/1204)

### CODE REVIEW

**Comment by Vampire** [comment]
The failing test is, because this does not work anymore in 3.0.5 which worked fine in 3.0.4:
grafik

**Comment by Vampire** [comment]
https://issues.apache.org/jira/browse/GROOVY-9665

**Comment by Vampire** [comment]
Actually this was an intended change.
And unfortunately rightfully even if a bit breaking for a minor release maybe.
I see two problematic cases if I didn't miss something.
`@IgnoreIf`/`@Requires`/`@PendingFeatureIf` (hereafter `@Requires` case) and `@Retry`.

For both we create the closure instances without `owner` and `thisObject`, with the context object as `delegate` and with `DELEGATE_ONLY` as resolve strategy.

For the `@Requires` case there is usually no instance of the specification class available yet, so to make the syntax work again, we would need to set the `owner` to the test class and the resolve strategy to `DELEGATE_FIRST`.

For the `@Retry` case we could even be fancier.
When this is evaluated, we have the current test instance available as this happens only after test failure.
So we could set the actual test instance as `owner` and `thisObject` and the resolve strategy to `DELEGATE_FIRST`.
This would enable the condition to reference the static fields like it was up to now and also the fields of the specification instance actually.

We could also make `@Retry` only be able to resolve the static fields like before with using the specification class as `owner` like we could for `@Requires`, effectively blocking access to the specification fields even if we could provide it.

Or actually we could make the `@Requires` cases more fancy and allow access to the specification fields there too.
Similar to how I made the data variables of an iteration available in the `@Requires` closures we could ues a similar strategy for specification fields, delaying the evaluation of the condition until the values are available.

So to summarize the multiple-choice options we have are:
- keep it like it is, if someone wants to use a static field, he has to qualify the access with `ClassName.FIELD`, no access to instance fields
- change the `@Requires` handling to allow simple access to static fields only
- change the `@Requires` handling to allow access to static fields and instance fields
- change the `@Retry` handling to allow simple access to static fields only
- change the `@Retry` handling to allow access to static fields and instance fields

@leonard84 & @szpak what do you think?

**Comment by leonard84** [comment]
IMHO we should only set the class as owner, doing it for the instance causes the setup phase to run, e.g., starting a spring context, only to abort again. And you'd either have to scan the whole class for valid properties/methods beforehand or you'd continue only to fail later anyway.

**Comment by Vampire** [comment]
In the `@Retry` case when the condition is evaluated, the instance is ready already anyway, as it is only evaluated after a failed execution of the feature.

For the `@Requires` case I would make it like with the data-variable access.
Evaluate the closure early like now without the instance available, and if failed due to a `MissingPropertyException`, then re-evaluate later when the instance actually is available using a feature or iteration interceptor.

Actually the logic is already present almost completely due to the shiny new data variable support.
I think the only missing parts would be to rehydrate the closure with the test instance as `owner` and `thisObject`, changing the resolve strategy to `DELEGATE_FIRST` and lifting the restriction a bit to fail on more or all `MissingPropertyException`s instead of only when a data variable name was tried.

**Comment by leonard84** [comment]
So you would not inspect the class to see if the missing property would even be available? Same with missing method?

**Comment by Vampire** [comment]
No, I think that would be unnecessary complex, superfluous, wasted time and too error prone.

If it fails due to missing property or method, then it could be that it was not a valid call to the test instance, then it will do the check again later and fail again. In that case this means it is a programming error in the test, so it has to be fixed to use a valid property or method. Once the test is fixed, it will only fail with missing property or missing method exception if an correct access was made. Checking that it was a valid access before checking again later would only be a waste of processing, besides that it is too error prone with the dynamic capabilities of Groovy.

For the data variables the check was easy and reliable, as we know the data variables that exist and there is not much magic that could happen additionally.

**Comment by leonard84** [comment]
Any other thoughts @spockframework/supporter

**Comment by leonard84** [comment]
If the implementation is analogous to `@Retry`, i.e., that the instance variables are available via `instance.var`, then I'm ok with it. Furthermore, we could reliably determine if the `MissingPropertyExceptions` was targeted at `instance`.

**Comment by henrik242** [comment]
(Latest groovy is 3.0.6, by the way)

**Comment by Vampire** [comment]
You are right @henrik242, but there is no reason to depend on a newer Groovy version if there is not something fixed that is essential for Spock imho.
3.0.5 fixes #1177, so it is better to depend on it.
@leonard84 what do you think, should we depend on the latest, or on the oldest compatible?

**Comment by szpak** [comment]
Fox me the newest version of Groovy the better. Usually there is more bugs fixed than new introduced ;). Some people just add Spock and have Groovy as a transitive dependency. Therefore, having that PR merged one day, I would also prefer 3.0.L(atest)

Nevertheless, it is rather more important to remember to do the dependency update right before the release. To keep it up-to-date.

Btw, if there is a suspicion that Spock depends on some Groovy feature added/fixed in some particular (minor) Groovy version there could be extra regression tests added. However, as I use them for my Gradle plugins, I'm definitely not convinced that Spock needs them.

File: docs/extensions.adoc [hunk 1]
```diff
@@ -117,6 +117,7 @@ To make predicates easier to read and write, the following properties are availa
 * `env` A map of all environment variables
 * `os` Information about the operating system (see `spock.util.environment.OperatingSystem`)
 * `jvm` Information about the JVM (see `spock.util.environment.Jvm`)
+* `instance` The specification instance if instance fields, shared fields or instance methods are needed
```
Comment (leonard84) [inline docs/extensions.adoc:120]: we should mention that instance will require the test to execute to that point, similar as with the data-driven variables.

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/ConditionalExtension.java [hunk 2]
```diff
@@ -108,7 +118,7 @@ public IterationCondition(Closure condition, T annotation) {
 
     @Override
     public void intercept(IMethodInvocation invocation) throws Throwable {
-      Object result = evaluateCondition(condition, invocation.getIteration().getDataVariables());
+      Object result = evaluateCondition(condition, invocation, invocation.getIteration().getDataVariables());
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/extension/builtin/ConditionalExtension.java:121]: Why do you pass the `invocation` and not the `invocation.getInstance()` here?

File: spock-core/src/main/java/org/spockframework/util/MopUtil.java [hunk 3]
```diff
@@ -25,8 +25,7 @@
 
 public abstract class MopUtil {
   private static final Field ReflectionMetaMethod_method  = getDeclaredField(ReflectionMetaMethod.class, "method");
-  private static final Field CachedField_field = GroovyRuntimeUtil.isGroovy2() ? getDeclaredField(CachedField.class, "field") :
-    getDeclaredField(CachedField.class, "cachedField");
+  private static final Field CachedField_field = getDeclaredField(CachedField.class, "field");
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/util/MopUtil.java:28]: was this fixed with groovy 3.0.5?

**Review by leonard84 — CHANGES_REQUESTED** [review]

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/ConditionalExtension.java (hunk 2 above)
Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/runtime/extension/builtin/ConditionalExtension.java:121]: Because it is done the same for `RetryConditionContext`, so I thought I just follow the same pattern here.
But I can of course also change that if you prefer.
If you do, should I change it for both?

File: spock-core/src/main/java/org/spockframework/util/MopUtil.java (hunk 3 above)
Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/util/MopUtil.java:28]: Depends on how you define "fixed".
Up to 3.0.5 it is `cachedField`.
In 3.0.6 they renamed it back to `field`.

**Review by leonard84 — APPROVED** [review]
